From edd4d36da2460dc7924c3f40ca2dee2d8ef583a9 Mon Sep 17 00:00:00 2001
From: David Faure <david.faure@kdab.com>
Date: Mon, 11 Aug 2025 15:45:08 +0200
Subject: [PATCH 241/553] qtestcase.h: fix -Wconversion warnings from
 std::chrono::duration vs int

Port more code to std::chrono to avoid this mix-n-match of precisions

formatTryTimeoutDebugMessage is exported but there's no BC requirement
in QTestLib, so this commit simply changes its signature.

Amends d4bb448cddce63e0c6a84a86020fa59dd32b2293.

Pick-to: 6.9 6.8
Change-Id: Idb9d0929d054fde308d22e5e128e11125591df4b
Reviewed-by: Ahmad Samir <a.samirh78@gmail.com>
(cherry picked from commit bca424b792fa38de6e5b38e4a9b17269ec2867ec)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/testlib/qtestcase.cpp |  6 ++++--
 src/testlib/qtestcase.h   | 15 +++++++++------
 2 files changed, 13 insertions(+), 8 deletions(-)

diff --git a/src/testlib/qtestcase.cpp b/src/testlib/qtestcase.cpp
index 6c58f327e13..5288998d2a5 100644
--- a/src/testlib/qtestcase.cpp
+++ b/src/testlib/qtestcase.cpp
@@ -375,11 +375,13 @@ void setThrowOnSkip(bool enable) noexcept
     g_throwOnSkip.fetchAndAddRelaxed(enable ? 1 : -1);
 }
 
-QString Internal::formatTryTimeoutDebugMessage(q_no_char8_t::QUtf8StringView expr, int timeout, int actual)
+QString Internal::formatTryTimeoutDebugMessage(q_no_char8_t::QUtf8StringView expr,
+                                               std::chrono::milliseconds timeout,
+                                               std::chrono::milliseconds actual)
 {
     return "QTestLib: This test case check (\"%1\") failed because the requested timeout (%2 ms) "
            "was too short, %3 ms would have been sufficient this time."_L1
-            .arg(expr, QString::number(timeout), QString::number(actual));
+            .arg(expr, QString::number(timeout.count()), QString::number(actual.count()));
 }
 
 extern Q_TESTLIB_EXPORT int lastMouseTimestamp;
diff --git a/src/testlib/qtestcase.h b/src/testlib/qtestcase.h
index c924f788106..42c601be9c6 100644
--- a/src/testlib/qtestcase.h
+++ b/src/testlib/qtestcase.h
@@ -175,7 +175,7 @@ inline void useVerifyThrowsException() {}
     if (!(expr)) { \
         QTest::qWait(0); \
     } \
-    int qt_test_i = 0; \
+    auto qt_test_i = std::chrono::milliseconds(0); \
     for (; qt_test_i < timeoutValue && !(QTest::runningTest() && QTest::currentTestResolved()) \
              && !(expr); qt_test_i += step) { \
         QTest::qWait(step); \
@@ -197,10 +197,11 @@ inline void useVerifyThrowsException() {}
             using namespace std::chrono_literals; \
             return std::chrono::milliseconds{timeoutAsGiven}; \
         }(); \
-    const int qt_test_step = qt_test_timeoutAsMs.count() < 350 ? qt_test_timeoutAsMs.count() / 7 + 1 : 50; \
-    const int qt_test_timeoutValue = qt_test_timeoutAsMs.count(); \
-    { QTRY_LOOP_IMPL(expr, qt_test_timeoutValue, qt_test_step) } \
-    QTRY_TIMEOUT_DEBUG_IMPL(expr, qt_test_timeoutValue, qt_test_step)
+    const auto qt_test_step = qt_test_timeoutAsMs < std::chrono::milliseconds(350) \
+                              ? qt_test_timeoutAsMs / 7 + std::chrono::milliseconds(1) \
+                              : std::chrono::milliseconds(50); \
+    { QTRY_LOOP_IMPL(expr, qt_test_timeoutAsMs, qt_test_step) } \
+    QTRY_TIMEOUT_DEBUG_IMPL(expr, qt_test_timeoutAsMs, qt_test_step)
 // Ends with an if-block, so doesn't want a following semicolon.
 
 // Will try to wait for the expression to become true while allowing event processing
@@ -341,7 +342,9 @@ namespace QTest
     Q_TESTLIB_EXPORT void maybeThrowOnSkip();
 
     Q_DECL_COLD_FUNCTION
-    Q_TESTLIB_EXPORT QString formatTryTimeoutDebugMessage(q_no_char8_t::QUtf8StringView expr, int timeout, int actual);
+    Q_TESTLIB_EXPORT QString formatTryTimeoutDebugMessage(q_no_char8_t::QUtf8StringView expr,
+                                                          std::chrono::milliseconds timeout,
+                                                          std::chrono::milliseconds actual);
     Q_TESTLIB_EXPORT Q_DECL_COLD_FUNCTION
     const char *formatPropertyTestHelperFailure(char *msg, size_t maxMsgLen,
                                                 const char *actual, const char *expected,
-- 
2.51.2

